Updates

PizzaCrust opened the pull request:
- Add `--recursive` to arguments in README.md.
- Update to the latest API changes in `MixinEntity`.

PizzaCrust committed:
Update Entity implementation

This commit updates and implements the MixinEntity class (the implementation and the mixin).

PizzaCrust committed:
Update README.md

stephan-gh commented on src/main/java/org/fountainmc/forge/mixin/entity/MixinEntity.java:
```diff
@@ -24,32 +24,159 @@
+        for (net.minecraft.entity.Entity entity : nmsEntities) {
+            apiEntities.add((Entity) entity);
+        }
+        return ImmutableList.copyOf(apiEntities);
```
Since Minecraft's `Entity` implements the interface you can directly do `ImmutableList.copyOf` on the `nmsEntities` for some performance boosts

PizzaCrust replied:
Doesn't work.

jamierocks replied:
^

PizzaCrust committed:
Directly cast to API entity

This commit casts the list and removes adding all and casting all items in the list.

jamierocks commented on src/main/java/org/fountainmc/forge/mixin/entity/MixinEntity.java:
```diff
@@ -24,32 +24,155 @@
+public abstract class MixinEntity implements Entity, ICommandSender {
 
     @Shadow public boolean onGround;
 
```
No gaps between fields

PizzaCrust committed:
Remove spaces from fields

jamierocks commented on src/main/java/org/fountainmc/forge/mixin/entity/MixinEntity.java:
```diff
@@ -24,32 +24,153 @@
+    public abstract List<net.minecraft.entity.Entity> shadow$getPassengers();
+
+    @Shadow
+    public abstract void shadow$addPassenger(net.minecraft.entity.Entity passenger);
```
Why the shadow$ prefix?

PizzaCrust replied:
IDE errors.

jamierocks commented on src/main/java/org/fountainmc/forge/mixin/entity/MixinEntity.java:
```diff
@@ -24,32 +24,153 @@
+
+    @Override
+    public void leaveVehicle() {
+        (this.getCommandSenderEntity()).dismountRidingEntity();
```
Why the brackets?

jamierocks commented on src/main/java/org/fountainmc/forge/mixin/entity/MixinEntity.java:
```diff
@@ -24,32 +24,153 @@
+    }
+
+    @Nonnull
+    public ImmutableList<Entity> entity$getPassengers() {
```
Why the entity$ if it's not intrinsic, or are you just missing intrinsic?

PizzaCrust committed:
Add intrinsic annotation to method

jamierocks commented on src/main/java/org/fountainmc/forge/mixin/entity/MixinEntity.java:
```diff
@@ -24,32 +24,151 @@
+    @Intrinsic
+    @Nonnull
+    public ImmutableList<Entity> entity$getPassengers() {
+        List entities = this.shadow$getPassengers();
```
Why the need for this field?

PizzaCrust replied:
IDE errors.

phase replied:
You sure get a lot of IDE errors.

PizzaCrust replied:
It doesn't work unless the shadow method has a List return type without a generic. Would Mixin register it correctly?

jamierocks replied:
Just cast to a List

PizzaCrust committed:
Formatted class

PizzaCrust committed:
Formatted class

jamierocks commented on src/main/java/org/fountainmc/forge/mixin/entity/MixinEntity.java:
```diff
@@ -24,32 +24,146 @@
+    @Override
+    public ImmutableCollection<Entity> getNearbyEntities(double distance) {
+        List<net.minecraft.entity.Entity> nmsEntities = this.getCommandSenderEntity().worldObj.getEntitiesWithinAABBExcludingEntity(this.getCommandSenderEntity(), this.getCommandSenderEntity().getEntityBoundingBox().expand(distance / 3, distance / 3, distance / 3));
+        return ImmutableList.copyOf(((List<Entity>)((List) nmsEntities)));
```
Don't bother casting to `List<Entity>`

jamierocks commented on src/main/java/org/fountainmc/forge/mixin/entity/MixinEntity.java:
```diff
@@ -24,32 +24,146 @@
+    @Intrinsic
+    @Nonnull
+    public ImmutableList<Entity> entity$getPassengers() {
+        return ImmutableList.copyOf((List<Entity>)((List) this.shadow$getPassengers()));
```
Don't bother casting to `List<Entity>`

jamierocks commented on src/main/java/org/fountainmc/forge/mixin/entity/MixinEntity.java:
```diff
@@ -24,32 +24,146 @@
 import org.fountainmc.api.world.World;
-import org.spongepowered.asm.mixin.Mixin;
-import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.*;
```
Don't collect imports using *.

jamierocks commented:
Status on this pr?
